lightning-block-sync/src/http_clients.rs

```diff
@@ -0,0 +1,1080 @@
+	/// format.
+	async fn post<F>(&mut self, uri: &str, host: &str, auth: &str, content: serde_json::Value) -> std::io::Result<F>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> {
+		let content = content.to_string();
+		write!(self.stream,
+			"POST {} HTTP/1.1\r\n\
+			 Host: {}\r\n\
+			 Authorization: {}\r\n\
+			 Connection: keep-alive\r\n\
+			 Content-Type: application/json\r\n\
+			 Content-Length: {}\r\n\
+			 \r\n\
+			 {}", uri, host, auth, content.len(), content)?;
+
+		let bytes = read_http_resp(&self.stream, MAX_HTTP_RESPONSE_LEN).await?;
+		F::try_from(bytes)
+	}
+}
+
+async fn read_http_resp(mut socket: &TcpStream, max_resp: usize) -> std::io::Result<Vec<u8>> {
+	let mut resp = Vec::new();
+	let mut bytes_read = 0;
+	macro_rules! read_socket { () => { {
+		#[cfg(feature = "tokio")]
+		let bytes_read = socket.read(&mut resp[bytes_read..]).await?;
+		#[cfg(not(feature = "tokio"))]
+		let bytes_read = socket.read(&mut resp[bytes_read..])?;
+		if bytes_read == 0 {
+			return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "zero bytes read"));
+		} else {
+			bytes_read
+		}
+	} } }
+
+	let mut actual_len = 0;
+	let mut ok_found = false;
+	let mut chunked = false;
+	// We expect the HTTP headers to fit in 8KB, and use resp as a temporary buffer for headers
+	// until we know our real length.
+	resp.extend_from_slice(&[0; 8192]);
+	'read_headers: loop {
+		if bytes_read >= 8192 {
+			return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "headers too large"));
+		}
+		bytes_read += read_socket!();
+		for line in resp[..bytes_read].split(|c| *c == '\n' as u8 || *c == '\r' as u8) {
+			let content_header = b"Content-Length: ";
+			if line.len() > content_header.len() && line[..content_header.len()].eq_ignore_ascii_case(content_header) {
+				actual_len = match match std::str::from_utf8(&line[content_header.len()..]){
+					Ok(s) => s, Err(e) => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, e)),
+				}.parse() {
+					Ok(len) => len, Err(e) => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, e)),
+				};
+			}
+			let http_resp_1 = b"HTTP/1.1 200 ";
+			let http_resp_0 = b"HTTP/1.0 200 ";
+			if line.len() > http_resp_1.len() && (line[..http_resp_1.len()].eq_ignore_ascii_case(http_resp_1) ||
+				                                  line[..http_resp_0.len()].eq_ignore_ascii_case(http_resp_0)) {
+				ok_found = true;
+			}
+			let transfer_encoding = b"Transfer-Encoding: ";
+			if line.len() > transfer_encoding.len() && line[..transfer_encoding.len()].eq_ignore_ascii_case(transfer_encoding) {
+				match &*String::from_utf8_lossy(&line[transfer_encoding.len()..]).to_ascii_lowercase() {
+					"chunked" => chunked = true,
+					_ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "unsupported transfer encoding")),
+				}
+			}
+		}
+		for (idx, window) in resp[..bytes_read].windows(4).enumerate() {
+			if window[0..2] == *b"\n\n" || window[0..2] == *b"\r\r" {
```

Comment:
this code will not be reached if the body is a single byte.  i.e. `\n\n5` is of length 3, so `windows(4)` will not stop the iteration just before we get to the end of the headers.